src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -185,8 +196,91 @@ public bool HasResponseStarted
             get { return _responseStarted; }
         }
 
+        private void TimeoutRequest()
+        {
+            if (TransitionToState(FrameState.Timeout) == FrameState.Timeout)
+            {
+                SocketInput?.AbortAwaiting();
+            }
+        }
+        protected int TransitionToState(int state)
+        {
+#pragma warning disable CS0420 // A reference to a volatile field will not be treated as volatile
```

Comment:
Yeah, my worry is that passing volatile fields as ref (or out) may have different behavior on different CPU architectures.